engine/runtime/src/main/java/org/enso/interpreter/runtime/data/atom/AtomConstructor.java

```diff
@@ -279,6 +387,10 @@ public String toString() {
    * @return the constructor function of this constructor.
    */
   public Function getConstructorFunction() {
+    if (constructorFunction == null) {
+      CompilerDirectives.transferToInterpreterAndInvalidate();
```

Comment:
Although, I'm not sure if it is necessary to invalidate and if it should be just a `transferToInterpreter()` call

Reply:
You certainly want to invalidate - the `constructorFunction == null` branch shall never get to generated machine code.

However I think there is better fix: Change the field:
```java
private @CompilerDirectives.CompilationFinal Supplier<Function> constructorFunction
```
to `CachingSupplier<Function>`. It has all the logic with `transferToInterpreterAndInvalidate()` already. Just _native image_ (I assume that was the reason for adding `@TruffleBoundary`) needs to know this cannot be _any_ `Supplier`, but the `final class CachingSupplier`.

Reply:
Unfortunately, providing the concrete supplier class does not fix the performance issue. On the branch [wip/db/register-type-definitions-overhead-1](https://github.com/enso-org/enso/tree/wip/db/10923-register-type-definitions-overhead-1) 52a4dc9d247b1ba995a678dcef54392a6fc78061 gives similar results as with the `@TruffleBoundary`

Reply:
That's hard to believe in. I'd like to see the IGV graph.

Reply:
The problem is that `AtomConstructor` isn't compilation constant itself. Then it makes no sense to use `@CompilationFinal` and yes, there is a difference between `Supplier<...>` and direct reference. 

![Not a compilation constant](https://github.com/user-attachments/assets/66b2aabd-d44d-40b3-98d9-6c61dabaffb7)

This might be a way to speed things up (even to current benchmark results):
```diff
enso$ git diff
diff --git engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeCallableNode.java engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeCallableNode.java
index d20afaed99..53727942ce 100644
--- engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeCallableNode.java
+++ engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeCallableNode.java
@@ -175,10 +175,14 @@ public abstract class InvokeCallableNode extends BaseNode {
     return this.invokeFunctionNode.execute(function, callerFrame, state, arguments);
   }
 
-  @Specialization
+  @Specialization(limit="1", guards = {
+      "constructor == cachedConstructor"
+  })
   Object invokeConstructor(
-      AtomConstructor constructor, VirtualFrame callerFrame, State state, Object[] arguments) {
-    return invokeFunction(constructor.getConstructorFunction(), callerFrame, state, arguments);
+      AtomConstructor constructor, VirtualFrame callerFrame, State state, Object[] arguments,
+      @Cached(value = "constructor", allowUncached = true) AtomConstructor cachedConstructor
+  ) {
+    return invokeFunction(cachedConstructor.getConstructorFunction(), callerFrame, state, arguments);
   }
 
   @Specialization
```
but that's probably out of scope of this PR. Feel free to continue with b94f81f if the results are on par.